src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -668,15 +693,39 @@ private static DkmEvaluationResult CreateEvaluationResultFromException(Exception
                 DataItem: null);
         }
 
-        internal static void ContinueWithExceptionHandling(CompletionRoutine onCompleted, CompletionRoutine<Exception> onException)
+        private sealed class WorkList
         {
-            try
+            internal readonly DkmWorkList InnerWorkList;
+            private readonly CompletionRoutine<Exception> _onException;
+            private CompletionRoutine _completionRoutine;
+
+            internal WorkList(DkmWorkList workList, CompletionRoutine<Exception> onException)
             {
-                onCompleted();
+                InnerWorkList = workList;
+                _onException = onException;
             }
-            catch (Exception e) when (ExpressionEvaluatorFatalError.ReportNonFatalException(e, DkmComponentManager.ReportCurrentNonFatalException))
+
+            internal void ContinueWith(CompletionRoutine completionRoutine)
             {
-                onException(e);
+                Debug.Assert(_completionRoutine == null);
+                _completionRoutine = completionRoutine;
+            }
+
+            internal void Execute()
+            {
+                while (_completionRoutine != null)
```

Comment:
_completionRoutine will be null if the item in progress has a [DebuggerDisplay] attribute since _completionRoutine will be set asynchronously.